xdsl/backend/riscv/lowering/convert_memref_to_riscv.py

```diff
@@ -76,60 +80,68 @@ def match_and_rewrite(self, op: memref.Dealloc, rewriter: PatternRewriter) -> No
-    element_type: Attribute,
+def get_strided_pointer(
+    src_ptr: SSAValue,
+    indices: Iterable[SSAValue],
+    memref_type: MemRefType[Any],
 ) -> tuple[list[Operation], SSAValue]:
     """
-    Returns ssa value representing pointer into the memref at given indices.
-    The pointer is byte-indexed, and the indices are strided by element size, so the index
-    into the flat memory buffer needs to be multiplied by the size of the element.
+    Given a buffer pointer 'src_ptr' which was originally of type 'memref_type', returns
+    a new pointer to the element being accessed by the 'indices'.
     """
-    assert len(shape) == len(indices)
 
-    bitwidth = bitwidth_of_type(element_type)
+    bitwidth = bitwidth_of_type(memref_type.element_type)
     if bitwidth % 8:
         raise DiagnosticException(
-            f"Cannot create offset for element type {element_type}"
+            f"Cannot create offset for element type {memref_type.element_type}"
             f" with bitwidth {bitwidth}"
         )
     bytes_per_element = bitwidth // 8
 
-    if not shape:
-        # Scalar memref
-        return ([], mem)
+    match memref_type.layout:
+        case NoneAttr():
+            strides = ShapedType.strides_for_shape(memref_type.get_shape())
+        case StridedLayoutAttr():
+            strides = memref_type.layout.get_strides()
+        case _:
+            raise DiagnosticException(f"Unsupported layout type {memref_type.layout}")
 
     ops: list[Operation] = []
 
-    head, *tail = indices
+    head: SSAValue | None = None
 
-    for factor, value in zip(shape[1:], tail):
-        if factor == 0:
-            continue
-        if factor == 1:
-            ops.extend(
-                (
-                    new_head_op := riscv.AddOp(
-                        head, value, rd=riscv.IntRegisterType.unallocated()
-                    ),
+    for index, stride in zip(indices, strides):
+        match stride:
+            case None:
+                raise NotImplementedError(
+                    f"MemRef {memref_type} with dynamic stride is not yet implemented"
                 )
-            )
-        else:
-            ops.extend(
-                (
-                    factor_op := riscv.LiOp(factor),
-                    offset_op := riscv.MulOp(
-                        factor_op.rd, head, rd=riscv.IntRegisterType.unallocated()
-                    ),
-                    new_head_op := riscv.AddOp(
-                        offset_op, value, rd=riscv.IntRegisterType.unallocated()
-                    ),
+            case 1:
+                pass
+            case _:
+                ops.extend(
+                    (
+                        stride_op := riscv.LiOp(stride),
+                        offset_op := riscv.MulOp(
+                            index, stride_op.rd, rd=riscv.IntRegisterType.unallocated()
+                        ),
+                    )
                 )
-            )
+                index = offset_op.rd
```

Comment:
!!

Reply:
Ok reading more tihs looks less like a bug than I initially thought but it might be worth a comment